packages/plugin-ext/src/plugin/plugin-context.ts

```diff
@@ -972,19 +972,33 @@ export function createAPIFactory(
     };
 }
 
-class Plugin<T> implements theia.Plugin<T> {
+enum ExtensionKind {
+    UI = 1,
+    Workspace = 2
+}
+
+export class PluginInstance<T> implements theia.Plugin<T> {
     id: string;
     pluginPath: string;
     pluginUri: theia.Uri;
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     packageJSON: any;
     pluginType: theia.PluginType;
+
     constructor(private readonly pluginManager: PluginManager, plugin: InternalPlugin) {
         this.id = plugin.model.id;
         this.pluginPath = plugin.pluginFolder;
         this.pluginUri = URI.parse(plugin.pluginUri);
         this.packageJSON = plugin.rawModel;
         this.pluginType = plugin.model.entryPoint.frontend ? 'frontend' : 'backend';
+
+        // Support VSCode Extension properties even if not explicitly exposed through interface
+        // as they are expected in the vscode.extension.* API calls and the ExtensionContext of the activate method
+        Object.assign(this, {
+            extensionPath: plugin.pluginPath,
+            extensionUri: plugin.pluginUri,
+            extensionKind: ExtensionKind.UI // stub as a local VS Code extension (not running on a remote workspace)
+        });
```

Comment:
It seems a bit odd to call `Object.assign` in the constructor. In this case, it's currently necessary because the fields added here don't appear in the class or interface declaration. There are a couple of options for addressing this:
 - Add these fields to the `theia.Plugin` interface.
 - Declare another interface that extends the `theia.Plugin` interface and have this class implement that interface.
 - Declare another interface and another class, but use that class by default rather than the previous `Plugin` class.

The last option is the most elaborate but causes the least breakage to downstream adopters who may have their own extensions of the `theia.Plugin` interface, but it may just defer problems if we eventually want to change our call signatures to require the new fields.

Reply:
You are right, I was just worried about the external visibility as I did not actually want to make it API. But I agree with you, it is probably inevitable and if we already have it we may as well define it in a well-structured manner. I followed your advice and implemented the third option to have a "base" Plugin class as before but an extension interface + implementation for the VS code-specific properties.